### PR TITLE
Feat: gguf with transformers

### PR DESCRIPTION
**Pull request by nikkolasg**

### CODE REVIEW

File: zkml/src/layers/transformer/mod.rs [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+            let op = concat_matmul::ConcatMatMul;
+            let qkt_v = op
+                .evaluate::<f32, GoldilocksExt2>(&vec![&qkt_reshaped, mha.outputs()[1]])?;
+            // now we some reshape/permute back 
+            // We go from [num_heads, 1, head_dim] → transpose back to [1, h, head_dim] → and reshape to [1, hidden_size]
+            let merged = qkt_v.outputs()[0].permute3d(&vec![1,0,2]).reshape(vec![1,self.hidden_size]);
+            // now we do the final projection
+            let projected = self.out.evaluate::<GoldilocksExt2>(&vec![&merged], vec![])?;
+            // and then residual connection
+            let out = input.add(&projected.outputs()[0]);
```
[the remark this answers is not in the document]
Comment (nicholas-mainardi) [reply zkml/src/layers/transformer/mod.rs:102]: Ah ok probably it's included in the feed-forward NN being executed after the transformer?

File: zkml/src/layers/transformer/mod.rs [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+    // Once this flat impl is consistent, then we can compare with the graph version.
+    // Once that is consistent too, we can delete.
+    struct FlatAttention<N> {
+        num_heads: usize,
+        head_dim: usize,
+        hidden_size: usize,
+        qkv: qkv::QKV<N>,
+        scaler: mul::ScalarMul<f32>,
+        reshape_q: reshape::Reshape,
+        reshape_kv: reshape::Reshape,
```
Comment (nicholas-mainardi) [inline zkml/src/layers/transformer/mod.rs:34]: These 2 reshapes look unused?

File: zkml/src/layers/transformer/mha.rs [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+
+impl MHA_QK {
+    pub fn new(num_heads: usize, head_dim: usize) -> Self {
+        Self {
+            num_heads,
+            head_dim,
+        }
+    }
+
+    pub fn evaluate<N: Number, E: ExtensionField>(
```
[the remark this answers is not in the document]
Comment (nikkolasg) [reply zkml/src/layers/transformer/mha.rs:31]: Yes - i made `ConcatMatMul` afterwards and i didn't change this one. 

> into multi-heads compatible matrices?

wdym ? 

> it looks weird that the layer is computing qk, but not the other operations of attention

A lot of things look weird because the proving is not done "by logical unit", e.g. MHA is actually (a) split heads & matmul (b) scalar division (c) softmax row by row and (d) matmul row by row.
All these operations we're proving separatedly. 
What do you suggest ?

File: zkml/src/layers/transformer/qkt.rs [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+    Tensor,
+    layers::{
+        LayerOut,
+        provable::{Evaluate, ProvableOpError},
+    },
+    tensor::Number,
+};
+
+#[derive(Clone, Debug)]
+pub struct QKT;
```
[the remark this answers is not in the document]
Comment (nikkolasg) [reply zkml/src/layers/transformer/qkt.rs:17]: Sure - happy to change. I just didn't have your matmul so i started everything from scratch. 
Yes it makes sense to integrate there as long as the API looks good. (the matmul might have a lot of config/options but happy to try!)

File: zkml/src/layers/transformer/mha.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (nikkolasg) [reply zkml/src/layers/transformer/mha.rs:31]: We discussed offline: the issue is quite complex to make sure proving doesn't need to re-do inference and there are multiple solutions, none without drawbacks. I took the simplest for me that makes it 1-1 with the proving logic but we need to modularize a bit more.
Once we have a clear proving logic implemented we can revisit.

File: zkml/src/layers/transformer/mod.rs [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+             // qk is now of shape [num_heads,seq_len]
+            assert_eq!(qkt_shape, vec![self.num_heads, seq_len]);
+            // We reshape to [num_heads, 1, seq_len] such concat_matmul can work, since it expects tensors of same shape
+            let qkt_reshaped = softmaxed.outputs()[0].clone().reshape(vec![self.num_heads, 1, seq_len]);
+            // now we can project back with V
+            let op = concat_matmul::ConcatMatMul;
+            let qkt_v = op
+                .evaluate::<f32, GoldilocksExt2>(&vec![&qkt_reshaped, mha.outputs()[1]])?;
+            // now we some reshape/permute back 
+            // We go from [num_heads, 1, head_dim] → transpose back to [1, h, head_dim] → and reshape to [1, hidden_size]
```
[the remark this answers is not in the document]
Comment (nikkolasg) [reply zkml/src/layers/transformer/mod.rs]: Well, it's actually one for now because i'm only considering one token, but when we treat a full sequence lenght, it's not gonna be 1 anymore.

File: zkml/assets/scripts/llms/gpt2_internal.py [hunk 6]
```diff
@@ -0,0 +1,373 @@
...
+model.eval()
+
+# Input
+input_text = "Hello"
+input_ids = tokenizer.encode(input_text, return_tensors="pt")
+
+# Run through model to get reference
+with torch.no_grad():
+    outputs = model(input_ids, output_hidden_states=True)
+    inputs_embeds = outputs.hidden_states[0] # This is wte(input_ids) + wpe(position_ids)
```
Comment (nicholas-mainardi) [inline zkml/assets/scripts/llms/gpt2_internal.py:43]: So basically here we are doing again all the layers in the model, saving outputs and compare that the final output is the same as the one of the imported GPT2 model?

File: zkml/src/parser/note.txt [hunk 7]
```diff
@@ -0,0 +1,454 @@
+
```
Comment (nicholas-mainardi) [inline zkml/src/parser/note.txt:1]: Was this added to the tracked files by mistake?

File: zkml/src/parser/json.rs [hunk 8]
```diff
@@ -0,0 +1,329 @@
...
+            .get_tensor("attn_qkv.bias")
+            .context("Failed to load attn_qkv.bias in from_json")?;
+
+        let hidden_size = c.hidden_size; // embedding_dim for GPT-2
+
+        // Unfuse weights:
+        // Expected shape of fused_qkv_weight is [3 * hidden_size, hidden_size] after python script transpose.
+        // Each individual q, k, v weight matrix should be [hidden_size, hidden_size].
+        // So, each chunk has hidden_size * hidden_size elements.
+        let weight_chunk_elements = hidden_size * hidden_size;
```
Comment (nicholas-mainardi) [inline zkml/src/parser/json.rs:94]: Shouldn't this be 'c.embedding_size * hidden_size`? The unfused tensors later are assumed to have this shape

File: zkml/src/layers/provable/error.rs [hunk 9]
```diff
@@ -0,0 +1,33 @@
+//! Contains the Error enum for things to do with an implementor of [`super::ProvableOp`].
+
+#[derive(thiserror::Error, Debug)]
+pub enum ProvableOpError {
```
Comment (nicholas-mainardi) [inline zkml/src/layers/provable/error.rs:4]: I understood we decided to remove custom errors? Why are we reintroducing them here?
